File: src/lib/code/mcs-216/session-1/2/2-horner.c
#include <stdio.h>

void evaluate_horner(int coeffs[], int n, int x) {
    long long result = coeffs[0];
    int mults = 0, adds = 0, iterations = 0;

    // The loop iterates from 1 to n-1 (total iterations = degree)
    for (int i = 1; i < n; i++) {
        iterations++;

        result = (result * x) + coeffs[i];

        mults++; // One multiplication per step
        adds++;  // One addition per step
    }

    printf("Result: %lld\n", result);
    printf("(a) Multiplications: %d, Additions: %d\n", mults, adds);
    printf("(b) Loop Iterations: %d\n", iterations);
}

int main() {
    int coeffs[] = {6, 5, 4, -3, 2, 8, -7};
    int n = sizeof(coeffs) / sizeof(coeffs[0]);
    evaluate_horner(coeffs, n, 3);
    return 0;
}

File: src/lib/code/mcs-216/session-1/2/2-horner.rs
struct Analysis {
    result: i64,
    multiplications: usize,
    additions: usize,
    iterations: usize,
}

fn horner_analysis(coeffs: &[i64], x: i64) -> Analysis {
    let mut res = coeffs[0];
    let mut m = 0;
    let mut a = 0;
    let mut i_count = 0;

    // Skipping the first element as it's our starting 'res'
    for &c in coeffs.iter().skip(1) {
        i_count += 1;
        res = (res * x) + c;
        m += 1;
        a += 1;
    }

    Analysis {
        result: res,
        multiplications: m,
        additions: a,
        iterations: i_count,
    }
}

fn main() {
    let coeffs = [6, 5, 4, -3, 2, 8, -7];
    let x = 3;
    let stats = horner_analysis(&coeffs, x);

    println!("Result: {}", stats.result);
    println!(
        "(a) Multiplications: {}, Additions: {}",
        stats.multiplications, stats.additions
    );
    println!("(b) Loop Iterations: {}", stats.iterations);
}
